aptos-move/block-executor/src/txn_last_input_output.rs

```diff
@@ -119,123 +124,168 @@ impl<T: Transaction, O: TransactionOutput<Txn = T>, E: Debug + Send + Clone>
-        )
-    }
-
-    pub(crate) fn check_fatal_vm_error(
+    // Should be called when txn_idx is committed, while holding commit lock.
+    //
+    // Records fee statement separately for block epilogue txn. This is done because the
+    // recorded output will be taken by materialization which can be concurrent with the
+    // block epilogue txn.
+    //
+    // Returns whether the block epilogue txn should be created. This is true when both
+    // of the following conditions hold:
+    // (1) the last txn in the block was committed (if any txns are left over, they must
+    // all be skipped), and
+    // (2) the last txn did not emit a new epoch event.
+    // To avoid unnecessarily inspecting events, we only check (2) if (1) is true.
+    pub(crate) fn commit<S: TStateView<Key = T::Key>>(
         &self,
         txn_idx: TxnIndex,
-    ) -> Result<(), ParallelBlockExecutionError> {
-        if let Some(status) = self.outputs[txn_idx as usize].load_full() {
-            if let ExecutionStatus::Abort(err) = status.as_ref() {
-                error!(
-                    "FatalVMError from parallel execution {:?} at txn {}",
-                    err, txn_idx
-                );
-                return Err(ParallelBlockExecutionError::FatalVMError);
-            }
-        }
-        Ok(())
-    }
-
-    pub(crate) fn check_execution_status_during_commit(
-        &self,
-        txn_idx: TxnIndex,
-    ) -> Result<(), PanicError> {
-        if let Some(status) = self.outputs[txn_idx as usize].load_full() {
-            match status.as_ref() {
-                ExecutionStatus::Success(_) | ExecutionStatus::SkipRest(_) => Ok(()),
+        num_txns: TxnIndex,
+        num_workers: usize,
+        user_txn_bytes_len: u64,
+        block_gas_limit_type: &BlockGasLimitType,
+        block_limit_processor: &mut BlockGasLimitProcessor<T, S>,
+        scheduler: &SchedulerWrapper,
+    ) -> Result<bool, PanicOr<ParallelBlockExecutionError>> {
+        let (mut skips_rest, mut must_create_epilogue_txn, maybe_fee_statement_and_output_size) =
+            match self.outputs[txn_idx as usize]
+                .load()
+                .as_ref()
+                .ok_or_else(|| {
+                    code_invariant_error(format!(
+                        "Execution output for txn {} not found during commit",
+                        txn_idx
+                    ))
+                })?
+                .as_ref()
+            {
+                ExecutionStatus::Success(output) => (
+                    false,
+                    (txn_idx == num_txns - 1 && !output.has_new_epoch_event_before_commit()?),
+                    Some((
+                        output.fee_statement_before_commit()?,
+                        output.output_approx_size()?,
+                    )),
+                ),
+                ExecutionStatus::SkipRest(output) => (
+                    true,
+                    !output.has_new_epoch_event_before_commit()?,
```

Comment:
@grao1991 if txn at index 1 is SkipRest, must_create_epilogue_txn here will be set to true if there is no new_epich_event_before_commit, and when commit returns prepare_and_queue_commit_ready_txns will also return true and the index will be set to 2.